File: CODING/SORTINGS/Quick_Sort.c
/* 
Sorting the elements of an Array using "QUICK SORT" 

Time Complexity : O(N ^ 2)
Space Complexity : O(N)
*/
#include<stdio.h>
#include<stdlib.h>

void PrintArray(int *arr, int size) {

	for(int i = 0; i < size; ++i) {
		printf(" %d,",arr[i]);
	}
}

void Swap(int *p, int *q) {

	int temp = *p;
	*p = *q;
	*q = temp;
}
/* This function takes last element as pivot, places 
the pivot element at its correct position in sorted 
array, and places all smaller (smaller than pivot) 
to left of pivot and all greater elements to right 
of pivot */
int Partition(int *arr, int low, int high) {

	int x = arr[high];
	int i = low - 1;

	for(int j = low; j < high; ++j) {
		// If current element is smaller than the pivot 
		if(arr[j] <= x) {
			Swap(&arr[++i], &arr[j]);
		}
	}
	Swap(&arr[++i], &arr[high]);
	return i;
}

void quickSort(int *arr, int low, int high) {

	if(low < high) {
        	/* pivot is partitioning index, arr[p] is now 
        	at right place */
		int pivot = Partition(arr, low, high);
        	/*Separately sort elements before 
        	partition and after partition */
		quickSort(arr, low, pivot - 1);
		quickSort(arr, pivot + 1, high);
	}
}

int main() {

	int *arr, size;

	printf("Enter size of the Array : ");	
	scanf("%d", &size);	

	arr = (int*)malloc(sizeof(int) * size);
	
	printf("Enter the Array : ");	
	for(int i = 0; i < size; ++i) {

		scanf("%d", &arr[i]);
	}

	PrintArray(arr, size);
	quickSort(arr, 0, size - 1);

	printf("** Sorted Successfully **");
	PrintArray(arr, size);
}

File: Core JAVA/1. Introduction/Programs with Recursions/Quick_sort.java
// Sorting using Quick sort
class Quick_sort
{
	public static int Partition(int arr[],int low,int high)
	{
		int x=arr[high];
		int i=low-1;
		for(int j=low;j<=high-1;j++)
		{
			if(arr[j]<=x)
			{
				i++;
				// Swapping (arr[i],arr[j]);
				int temp=arr[i];
				arr[i]=arr[j];
				arr[j]=temp;
			}
		}
		// Swapping (arr[i+1],arr[high]);
		int temp=arr[i+1];
		arr[i+1]=arr[high];
		arr[high]=temp;
		return (i+1);
	}
	public static void Sort(int arr[],int low,int high)
	{
		if(low<high)
		{
			int pivot=Partition(arr,low,high);
			Sort(arr,low,pivot-1);
			Sort(arr,pivot+1,high);
		}
	}
	public static void main(String args[])
	{
		int arr[]={3,7,1,5,6,10,11,34,21,56,33};
		Quick_sort ob=new Quick_sort();
		int n=arr.length-1;
		System.out.println("\n The intial array : ");
		for(int i=0;i<=n;i++)
		{
			System.out.print("  "+arr[i]);
		}
		ob.Sort(arr,0,n);
		System.out.println("\n The sorted array : ");
		for(int i=0;i<=n;i++)
		{
			System.out.print("  "+arr[i]);
		}
		System.out.println();
	}
}